### PR TITLE
Add a per-source floor for Auto Granularity

### PR DESCRIPTION
**Pull request by arj22**
## What this does

"Auto Granularity" picks a chart's bucket size based only on the selected time range — it has no idea how often the underlying metric actually reports. On a short range, that can pick a bucket smaller than the metric's real scrape interval, so the chart looks sparse/steppy (alternating real values and gaps) even though nothing's actually wrong with the data.

This adds a **Minimum Auto Granularity** setting on a Metric source (Team Settings → Sources → your Metrics source) that puts a floor under that auto-inference. It's the same idea as Grafana's per-datasource "Min interval" setting. It only affects "Auto" — a tile with an explicit granularity picked is untouched.

The following happens on 5m, 15m and 30m durations on one of our tiles. Similar issues on many tiles. 

Screenshot 2026-09-18 at 12 40 06 AM

## Scope

Covers dashboards/tiles only. Alerts, the external API, the CLI, and the MCP query tools build their queries a different way and don't get this floor yet — didn't want to widen the diff for surfaces I wasn't using.

## Testing

New unit tests for the floor logic, full existing test suite passes, typechecks clean. Not manually clicked through in a browser yet — flagging that for review rather than claiming it works end-to-end.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx [hunk 1]
```diff
@@ -21,11 +22,33 @@ import { DEFAULT_DATABASE, OTEL_CLICKHOUSE_EXPRESSIONS } from './constants';
...
   control,
   setValue,
   sourceId,
 }: TableModelProps) {
+  // Metric-only field, not present on every TSource union member (same
+  // reason DBTableSelectControlled/SourceSelectControlled below type their
+  // own `name` prop loosely rather than as `Path<TSource>`).
+  const { field: minAutoGranularityField } = useController({
+    control,
+    name: 'minAutoGranularity' as any,
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx]: Fixed — replaced the hand-rolled useController + Select with the project's existing SelectControlled component, which removes the `as any` cast entirely.

File: packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx [hunk 2]
```diff
@@ -320,6 +343,16 @@ export function MetricTableModelForm({
         >
           <SourceSelectControlled control={control} name="logSourceId" />
         </FormRow>
+        <FormRow
+          label="Minimum Auto Granularity"
+          helpText="Floor for 'Auto Granularity' on charts querying this source. Set this to your metrics' scrape/report interval to avoid sparse-looking charts on short time ranges. Doesn't affect an explicitly chosen (non-Auto) granularity."
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx]: Fixed — label/help text are now sentence case, matching MV_GRANULARITY_OPTIONS's own style since the option list now comes directly from there.

File: packages/app/src/hooks/useChartConfig.tsx [hunk 3]
```diff
@@ -428,7 +445,10 @@ export function useQueriedChartConfig(
         };
       }
 
-      const optimizedConfig = mvOptimizationData?.optimizedConfig ?? config;
+      const optimizedConfig = {
+        ...(mvOptimizationData?.optimizedConfig ?? config),
+        minGranularitySeconds: getMinGranularitySeconds(source),
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/hooks/useChartConfig.tsx]: Confirmed and fixed. Threaded minGranularitySeconds through getTimeChartGranularity/convertToTimeChartConfig/useTimeChartSettings instead, and resolved the source (and its floor) earlier — before conversion — at each real call site (DBTimeChart, NumberTileBackgroundChart, SearchTotalCountChart, and the chart-editor's buildChartConfigForExplanations). Verified with the existing DBTimeChart/NumberTileBackgroundChart/DBSearchPageQueryKey test suites (all pass) plus new unit tests.

File: packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx [hunk 4]
```diff
@@ -21,11 +22,33 @@ import { DEFAULT_DATABASE, OTEL_CLICKHOUSE_EXPRESSIONS } from './constants';
...
+// Deliberately excludes 'auto' (this floors auto-inference, it doesn't
+// replace it) and stops at 1 hour - a source reported that infrequently is
+// past the point where a fixed minimum bucket size is a sane default.
+const MIN_AUTO_GRANULARITY_OPTIONS = [
+  { value: '', label: 'No minimum' },
+  { value: Granularity.FifteenSecond, label: '15 Seconds' },
+  { value: Granularity.ThirtySecond, label: '30 Seconds' },
+  { value: Granularity.OneMinute, label: '1 Minute' },
+  { value: Granularity.FiveMinute, label: '5 Minutes' },
+  { value: Granularity.TenMinute, label: '10 Minutes' },
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx]: Fixed — dropped the hand-rolled ladder and now derive the option list from the existing MV_GRANULARITY_OPTIONS, which is already curated to match exactly what convertDateRangeToGranularityString can return.

File: packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx [hunk 5]
```diff
@@ -21,11 +22,33 @@ import { DEFAULT_DATABASE, OTEL_CLICKHOUSE_EXPRESSIONS } from './constants';
...
+
 export function MetricTableModelForm({
   control,
   setValue,
   sourceId,
 }: TableModelProps) {
+  // Metric-only field, not present on every TSource union member (same
+  // reason DBTableSelectControlled/SourceSelectControlled below type their
+  // own `name` prop loosely rather than as `Path<TSource>`).
+  const { field: minAutoGranularityField } = useController({
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx]: Fixed — switched to SelectControlled as suggested, and moved the option list into constants.ts alongside USE_TEXT_INDEX_OPTIONS.

File: packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx [hunk 6]
```diff
@@ -320,6 +343,16 @@ export function MetricTableModelForm({
         >
           <SourceSelectControlled control={control} name="logSourceId" />
         </FormRow>
+        <FormRow
+          label="Minimum Auto Granularity"
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx]: Fixed — sentence case throughout.

File: packages/common-utils/src/__tests__/utils.test.ts [hunk 7]
```diff
@@ -3303,4 +3304,44 @@ describe('utils', () => {
       ).toBe('EventTime');
     });
   });
+
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/common-utils/src/__tests__/utils.test.ts:3307]: Added both: a source persistence round-trip test in sources.int.test.ts (mirroring the seriesTable block, including the ''-coerces-to-unset case), and a getMinGranularitySeconds unit test that starts from a metric TSource with minAutoGranularity set and asserts the converted seconds value — the actual source-to-floor conversion this feature depends on, not just the pure utility function with the floor handed in directly.

File: packages/app/src/components/DBTimeChart.tsx [hunk 8]
```diff
@@ -399,19 +402,28 @@ function DBTimeChartComponent({
...
+  const minGranularitySeconds = getMinGranularitySeconds(source);
+
   const originalDateRange = config.dateRange;
   const {
     displayType: displayTypeProp,
     dateRange,
     granularity,
     fillNulls,
-  } = useTimeChartSettings(config);
+  } = useTimeChartSettings({ ...config, minGranularitySeconds });
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/DBTimeChart.tsx]: Confirmed and fixed — hoisted the merge into a single useMemo (configWithFloor) used by both useTimeChartSettings and convertToTimeChartConfig, instead of an inline spread at each call site.

File: packages/app/src/hooks/useChartConfig.tsx [hunk 9]
```diff
@@ -428,7 +455,10 @@ export function useQueriedChartConfig(
         };
       }
 
-      const optimizedConfig = mvOptimizationData?.optimizedConfig ?? config;
+      const optimizedConfig = {
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/hooks/useChartConfig.tsx:452]: Fixed — minGranularitySeconds is now computed above useQuery and included in the queryKey for both useQueriedChartConfig and useRenderedSqlChartConfig.

File: packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx [hunk 10]
```diff
@@ -21,11 +22,28 @@ import { DEFAULT_DATABASE, OTEL_CLICKHOUSE_EXPRESSIONS } from './constants';
 import { FormRow } from './FormRow';
 import { TableModelProps } from './types';
 
+// Reuses MV_GRANULARITY_OPTIONS (rather than restating the ladder here)
+// because it's already curated to match what convertDateRangeToGranularityString
+// can actually return (e.g. it omits '10 minute', which the auto-inference
+// algorithm skips in favor of '15 minute' - a hand-rolled list that included
+// it would silently floor a "10 minute" choice to 15 minutes instead).
+const MIN_AUTO_GRANULARITY_OPTIONS = [
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx]: Fixed — filtered '1 second' out of the reused MV_GRANULARITY_OPTIONS list.

File: packages/app/src/hooks/useChartConfig.tsx [hunk 11]
```diff
@@ -96,6 +99,26 @@ const shouldUseChunking = (
...
+ *
+ * Exported because `'auto'` is resolved to a concrete granularity by
+ * `convertToTimeChartConfig`/`useTimeChartSettings` (ChartUtils.tsx) before a
+ * time-chart's config ever reaches `useQueriedChartConfig` below - callers
+ * that build a chart config outside of this hook (DBTimeChart and its
+ * siblings) need to compute the same floor and apply it earlier, or the
+ * floor set here never takes effect (granularity is no longer `'auto'` by
+ * the time this hook's queryFn runs).
+ */
+export function getMinGranularitySeconds(
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/hooks/useChartConfig.tsx:109]: Added two DBTimeChart tests asserting the floor actually reaches the resolved granularity end-to-end (a 15-minute range that would otherwise infer '15 second' correctly floors to '1 minute' when the source's minAutoGranularity is set, and is unaffected when it isn't) - not just the pure conversion functions in isolation.

File: packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx [hunk 12]
```diff
@@ -320,6 +338,16 @@ export function MetricTableModelForm({
         >
           <SourceSelectControlled control={control} name="logSourceId" />
         </FormRow>
+        <FormRow
+          label="Minimum Auto Granularity"
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx]: This one looks like it was generated against an earlier revision - by the time it posted, MetricTableModelForm.tsx already read `label="Minimum auto granularity"` and `helpText="Floor for 'auto granularity' on charts querying this source. ... Doesn't affect an explicitly chosen (non-auto) granularity."`, both sentence case. No further change needed.

File: packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx [hunk 13]
```diff
@@ -320,6 +325,17 @@ export function MetricTableModelForm({
...
+        <FormRow
+          label="Minimum auto granularity"
+          helpText="Floor for 'auto granularity' on charts querying this source. Set this to your metrics' scrape/report interval to avoid sparse-looking charts on short time ranges. Doesn't affect an explicitly chosen (non-auto) granularity."
+        >
+          <SelectControlled
+            control={control}
+            name="minAutoGranularity"
+            data={MIN_AUTO_GRANULARITY_OPTIONS}
+            allowDeselect={false}
+          />
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/MetricTableModelForm.tsx:343]: Fixed — added placeholder="No minimum" to the SelectControlled. Also found (via actually running the integration test, not flagged by review) that the empty-string value wasn't being coerced to unset at all on save: this router's validateRequest middleware validates req.body against the Zod schema but never writes the parsed/transformed result back to it, so the schema's own '' -> undefined preprocess never ran. Added a Mongoose setter on the field itself, which is the one place every write path actually funnels through.

File: packages/app/src/hooks/useChartConfig.tsx [hunk 14]
```diff
@@ -428,7 +455,10 @@ export function useQueriedChartConfig(
         };
       }
 
-      const optimizedConfig = mvOptimizationData?.optimizedConfig ?? config;
+      const optimizedConfig = {
+        ...(mvOptimizationData?.optimizedConfig ?? config),
+        minGranularitySeconds: getMinGranularitySeconds(source),
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/hooks/useChartConfig.tsx]: This was already fixed in the prior commit (db0a0896/4d53f11c) - minGranularitySeconds has been part of both query keys (queryKey: [config, ..., minGranularitySeconds] and ['renderedSql', config, minGranularitySeconds]) since before this comment posted. Looks like it was generated against an earlier revision.

File: packages/app/src/components/__tests__/DBTimeChart.test.tsx [hunk 15]
```diff
@@ -20,6 +20,7 @@ import { useSource } from '@/source';
 // Mock dependencies
 jest.mock('@/hooks/useChartConfig', () => ({
   useQueriedChartConfig: jest.fn(),
+  getMinGranularitySeconds: jest.fn().mockReturnValue(undefined),
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/__tests__/DBTimeChart.test.tsx:26]: Already fixed before this comment posted - DBTimeChart.test.tsx has two tests asserting the floor reaches the resolved granularity end-to-end (with getMinGranularitySeconds mocked to return 60, and with it returning undefined), added in commit 4d53f11c.

File: packages/app/src/components/Sources/SourceForm/constants.ts [hunk 16]
```diff
@@ -24,6 +27,16 @@ export const OTEL_CLICKHOUSE_EXPRESSIONS = {
   resourceAttributesExpression: 'ResourceAttributes',
 };
 
+// Reuses MV_GRANULARITY_OPTIONS (rather than restating the ladder here)
+// because it's already curated to match what convertDateRangeToGranularityString
+// can actually return (e.g. it omits '10 minute', which the auto-inference
+// algorithm skips in favor of '15 minute' - a hand-rolled list that included
+// it would silently floor a "10 minute" choice to 15 minutes instead).
+export const MIN_AUTO_GRANULARITY_OPTIONS = [
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/Sources/SourceForm/constants.ts:33]: Already fixed before this comment posted - MIN_AUTO_GRANULARITY_OPTIONS filters out '1 second' (constants.ts), added in commit 4d53f11c.

File: packages/app/src/hooks/__tests__/useChartConfig.test.tsx [hunk 17]
```diff
@@ -1603,6 +1606,56 @@ describe('useChartConfig', () => {
...
+      expect(
+        getMinGranularitySeconds({
+          id: 'source-2',
+          kind: SourceKind.Log,
+          name: 'Test Logs',
+          connection: 'conn-1',
+          from: { databaseName: 'default', tableName: 'otel_logs' },
+          timestampValueExpression: 'TimestampTime',
+          defaultTableSelectExpression: 'Body',
+        } as TSource),
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/hooks/__tests__/useChartConfig.test.tsx]: Fixed — switched to satisfies TSource.

File: packages/app/src/components/DBEditTimeChartForm/utils.ts [hunk 18]
```diff
@@ -385,7 +386,10 @@ export function buildChartConfigForExplanations({
   // other at runtime, so the SQL preview transforms `config` itself into
   // both queries on render and the MV indicator is suppressed for this
   // tab.  Returning `config` unchanged is intentional.
-  const builderConfig = config as BuilderChartConfigWithDateRange;
+  const builderConfig: BuilderChartConfigWithDateRange = {
```
[the remark this answers is not in the document]
Comment (arj22) [reply packages/app/src/components/DBEditTimeChartForm/utils.ts:389]: Fixed — removed the cast. The preceding isBuilderChartConfig(config) guard already narrows config enough for the object literal to satisfy BuilderChartConfigWithDateRange without it.

**Comment by arj22** [comment]
@jordan-simonovski can you review this? if someone else has to review, please assign them 🙏

File: .changeset/source-min-auto-granularity.md [hunk 19]
```diff
@@ -0,0 +1,6 @@
+---
+'@hyperdx/common-utils': minor
+'@hyperdx/app': minor
```
Comment (jordan-simonovski) [inline .changeset/source-min-auto-granularity.md:3]: **chore (non-blocking):** add `@hyperdx/api` to the changeset?

`models/source.ts` gained a persisted field. The `fixed` group means the version bump
lands either way, so this is only about the api changelog naming the change.

File: packages/app/src/components/DBTimeChart.tsx [hunk 20]
```diff
@@ -399,19 +402,38 @@ function DBTimeChartComponent({
     [],
   );
 
+  // Resolved before useTimeChartSettings/convertToTimeChartConfig below (not
```
Comment (jordan-simonovski) [inline packages/app/src/components/DBTimeChart.tsx]: **suggestion (non-blocking):** could most of the new comments come out?

`AGENTS.md` sets a budget of one comment line per five lines of code and this diff runs
the other way: 13 of the 26 added lines here, 8 of 12 in `SourceForm/constants.ts`,
9 of 13 in the api `source.ts`. Most of it is the investigation story - why the hook
moved, what a previous review found, what `validateRequest` does instead of
`processRequest`, which is a bit redundant. We could just go with:

```tsx
// Both useTimeChartSettings and convertToTimeChartConfig resolve 'auto', so the
// minimum has to be in `config` before they run.
```

**Review by jordan-simonovski — DISMISSED** [review]
Looking pretty good apart from a couple of suggestions. Thanks for the contribution!

File: .changeset/source-min-auto-granularity.md (hunk 19 above)
Comment (arj22) [reply .changeset/source-min-auto-granularity.md:3]: Good catch, added.

File: packages/app/src/components/DBTimeChart.tsx (hunk 20 above)
Comment (arj22) [reply packages/app/src/components/DBTimeChart.tsx]: Fair - trimmed these down (and a couple others in the same vein: useChartConfig.tsx, NumberTileBackgroundChart.tsx, SearchTotalCountChart.tsx) to the load-bearing why, dropped the investigation narration.

File: .changeset/source-min-auto-granularity.md [hunk 21]
```diff
@@ -0,0 +1,7 @@
...
+'@hyperdx/api': patch
+---
+
+Add a per-source floor for "Auto Granularity". A metric source can now set "Minimum Auto Granularity" (Team Settings → Sources → your Metrics source) so that auto-inferred time buckets never go below it — useful when the underlying metric is reported on a fixed interval (e.g. a 60s scrape), since a short selected date range can otherwise auto-infer a smaller bucket than that interval and render a sparse/steppy series (alternating real-sample/empty buckets). Mirrors Grafana's per-datasource "Min interval" setting. Unset (the default) preserves the existing unfloored behavior, and an explicit (non-"Auto") granularity chosen on a tile is never affected.
```
[the remark this answers is not in the document]
Comment (arj22) [reply .changeset/source-min-auto-granularity.md]: Fixed — sentence case throughout.

**Comment by arj22** [comment]
@jordan-simonovski thanks. this is ready to be merged

**Comment by brandon-pereira** [comment]
@arj22 can we please fix/check this comment?

> 🟠 packages/api/src/mcp/tools/sources/schemas.ts:164 — MCP clickstack_save_source silently wipes minAutoGranularity on every metric-source update → mcpSaveSourceSchema is a hand-maintained flat mirror of the source shape and has no minAutoGranularity entry, so describe_source (which derives its round-trippable config from Object.keys(mcpSaveSourceSchema.shape) in extractSourceConfig) never emits it and save_source can never receive it. Because updateSource in packages/api/src/controllers/sources.ts:84 is a findOneAndReplace, any MCP read-modify-write of a metric source — e.g. an agent renaming the source — drops the user's setting with no warning. Add minAutoGranularity: z.string().optional() under the ── Metric ── section (next to metricTables) so it round-trips like the other metric fields.

**Comment by arj22** [comment]
@brandon-pereira Fixed in 001f4041 — added `minAutoGranularity` to the MCP save schema and a test proving it survives a rename now. Ran the full sources test suite against a real DB, all passing.

**Review by brandon-pereira — APPROVED** [review]
LGTM - thanks for the contribution

**Comment by arj22** [comment]
@brandon-pereira  Can you approve the CI workflows?
